dask/dataframe/io/parquet/arrow.py

```diff
@@ -1646,14 +1473,21 @@ def _read_table(
             # Check if we need to generate a fragment for filtering.
             # We only need to do this if we are applying filters to
             # columns that were not already filtered by "partition".
-            if partitioning and _need_fragments(filters, partition_keys):
+            #
+            # Note that we also generate a new fragment if the user
+            # has passed in an explicit `arrow_fs` object.
+            if (
+                (partitions and partition_keys is None)
+                or arrow_filesystem
+                or (partitioning and _need_fragments(filters, partition_keys))
+            ):
 
                 # We are filtering with "pyarrow-dataset".
                 # Need to convert the path and row-group IDs
                 # to a single "fragment" to read
                 ds = pa_ds.dataset(
                     path_or_frag,
-                    filesystem=fs,
+                    filesystem=arrow_filesystem or fs,
```

Comment:
@jrbourbeau @martindurant - Note that I added this `arrow_filesystem` option for `engine="pyarrow-dataset"`, because I noticed that pyarrow is significantly faster when it is working with a native file-object type. When this option is used, the pyarrow object is only used at IO time (not for metadata collection/processing, where the fsspec API is required).  Some motivation:

```python
import dask.dataframe as dd
from pyarrow import fs as pa_fs

# Large parquet file with 100 columns
path = "s3://<my-bucket>/large-file.parquet"

# Using pyarrow-filesystem object
%%time
ddf = dd.read_parquet(
    path,
    engine="pyarrow-dataset",
    arrow_filesystem=pa_fs.FileSystem.from_uri(path)[0],
)
len(ddf["id10"])
CPU times: user 663 ms, sys: 238 ms, total: 901 ms
Wall time: 10 s

# Using default s3fs object
%%time
ddf = dd.read_parquet(
    path,
    engine="pyarrow-dataset",
)
len(ddf["id10"])
CPU times: user 7.32 s, sys: 1.25 s, total: 8.57 s
Wall time: 1min 59s
```

Note that it would be nice to handle the arrow-filesystem object construction automatically, but I think it makes sense to avoid making any strong "promises" in this PR.  This `arrow_filesystem` hook is intended to enable further experimentation, without changing default behavior (for now).  Any thoughts on this?

Reply:
Obviously, I would appreciate if you could investigate why s3fs is slower - is it making the same requests?

Reply:
>Obviously, I would appreciate if you could investigate why s3fs is slower - is it making the same requests?


I'm certainly interested to know why the difference is so large as well.  Especially since there is no arrow-backed c++ file-system to use for gcs.

According to pyarrow documentation, [an fsspec filesystem is converted to a `PyFileSystem` object](https://arrow.apache.org/docs/python/filesystems.html#using-fsspec-compatible-filesystems).  The individual files are ultimately converted into [`PythonFile` instances](https://github.com/apache/arrow/blob/master/python/pyarrow/io.pxi#L676), which wrap  [`PyReadableFile` objects](https://github.com/apache/arrow/blob/8e43f23dcc6a9e630516228f110c48b64d13cec6/cpp/src/arrow/python/io.h#L33) (which are `RandomAccessFile` instances).  The pyarrow code does include [comments/warnings](https://github.com/apache/arrow/blob/master/python/pyarrow/io.pxi#L683-L685) that using python-based file objects is expected to be slow.

As far as I can tell, the parquet is implemented with `ReadAt` calls.  For the fsspec-backed files, this seems to be a [seek + read operation](https://github.com/apache/arrow/blob/8e43f23dcc6a9e630516228f110c48b64d13cec6/cpp/src/arrow/python/io.cc#L252) (would `read_block` be any faster here?).  For the arrow-native s3 file-object, [the logic is a bit more specialized](https://github.com/apache/arrow/blob/d5a2aa2ffb1c2fc4f3ca48c829fcdba80ec67916/cpp/src/arrow/filesystem/s3fs.cc#L909).

Therefore, it seems like the arrow-native version is faster because (1) there is no python overhead, and (2) the specific requests used to read from the file may be better optimized for s3.

Reply:
Do you have this specific dataset at a public place so I can have a look?

Reply:
Meanwhile, cross-posting from slack, about the fsspec change that will certainly fix this:
> For an fsspec function cat(list-of-paths, start=list-of-start, end=list-of-end), what would be a reasonable return? Currently cat returns either bytes for a single path, or {path: bytes} for more than one path.

Reply:
Note that I rolled back the `arrow_filesystem` change, because the performance is much better when the `cache_type` is set to "none" or "all" (depending on the type of read) in `fs.open`.  Note that we can include the `cat_ranges` optimization in a follow-up PR.